src/main/java/org/broadinstitute/hellbender/tools/spark/BaseRecalibratorSparkOptimized.java

```diff
@@ -0,0 +1,205 @@
+import org.broadinstitute.hellbender.utils.variant.Variant;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@CommandLineProgramProperties(
+        summary = "Base Quality Score Recalibration (BQSR) -- Generates recalibration table based on various user-specified covariates (such as read group, reported quality score, machine cycle, and nucleotide context).",
+        oneLineSummary = "Generates recalibration table",
+        programGroup = SparkProgramGroup.class
+)
+public class BaseRecalibratorSparkOptimized extends SparkCommandLineProgram {
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * all the command line arguments for BQSR and its covariates
+     */
+    @ArgumentCollection(doc = "all the command line arguments for BQSR and its covariates")
+    private final RecalibrationArgumentCollection bqsrArgs = new RecalibrationArgumentCollection();
+
+    // inputs can be local, GCS, or HDFS.
+    @ArgumentCollection
+    private final RequiredReadInputArgumentCollection readArguments = new RequiredReadInputArgumentCollection();
+
+    @ArgumentCollection
+    private final IntervalArgumentCollection intervalArgumentCollection = new OptionalIntervalArgumentCollection();
+
+    @ArgumentCollection
+    private final ReferenceInputArgumentCollection referenceArguments = new RequiredReferenceInputArgumentCollection();
+
+    @Argument(doc = "the known variants. Must be local.", shortName = "knownSites", fullName = "knownSites", optional = false)
+    private List<String> knownVariants;
+
+    // output can be local or GCS.
+    @Argument(doc = "Path to save the final recalibration tables to.",
+              shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    private String outputTablesPath = null;
+
+
+    // the granularity at which we'll want to assign work and read inputs.
+    private final int bigShardSize = 1_000_000;
+    // the granularity at which we want to batch reads,variants,and reference for processing.
+    private final int smallShardSize = 5_000;
+    // we make the assumption that all reads that start in a shard
+    // will end within "margin" of the shard. This allows us to split the computation
+    // across machines.
+    private final int margin = 1000;
+
+    private AuthHolder auth;
+
+    /**
+     * Reference window function for BQSR. For each read, returns an interval representing the span of
+     * reference bases required by the BQSR algorithm for that read. Should be passed into the
+     * {@link ReferenceMultiSource} object for the {@link AddContextDataToRead} transform.
+     */
+    public static final SerializableFunction<GATKRead, SimpleInterval> BQSR_REFERENCE_WINDOW_FUNCTION =
+            read -> BAQ.getReferenceWindowForRead(read, BAQ.DEFAULT_BANDWIDTH, BAQ.DEFAULT_INCLUDE_CLIPPED_BASES);
```

Comment:
Delete, and use `BaseRecalibrationEngine.BQSR_REFERENCE_WINDOW_FUNCTION` instead.